### PR TITLE
Feature/99 robustness utility

### PR DESCRIPTION
**Pull request by kasra-hosseini**
Main changes in this PR:
* Generate a report (in HTML format) based on the results of `sklearn_classifiers.py`
* Compute confusion matrix 
* Plot confusion matrix (this function is used when generating a report)
* Functions to collect and print `sklearn_classifier` outputs
* fillna for both original and synthetic datasets

Example report:

Screenshot 2020-05-18 at 15 37 43

And for confusion matrices:

Screenshot 2020-05-18 at 15 36 52

Details:

Screenshot 2020-05-18 at 15 37 53

### CODE REVIEW

File: run-inputs/not_used/polish-synthpop-dr-4.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    },
+    "parameters_sklearn_utility":
+    {
+        "enabled": true,
+        "input_columns": ["sex", "age", "edu", "marital", "sport", "ls", "smoke"],
+        "label_column": "wkabint",
+        "test_train_ratio": 0.3,
+        "num_leaked_rows": 0
+    }
+}
```
Comment (gmingas) [inline run-inputs/not_used/polish-synthpop-dr-4.json]: Should we remove this not_used folder from develop?

Also, this has conflicts with #38 as that branch contains the extra key for correlations.

File: run-inputs/polish-synthpop-dr-4.json [hunk 2]
```diff
@@ -9,24 +9,24 @@
...
-        "input_columns": ["sex", "age", "edu", "marital", "sport", "ls", "smoke"],
-        "label_column": "wkabint",
+        "input_columns": ["sex", "wkabint", "marital", "sport", "ls", "smoke"],
+        "label_column": "edu",
         "test_train_ratio": 0.3,
         "num_leaked_rows": 0
     }
-}
\ No newline at end of file
+}
```
Comment (gmingas) [inline run-inputs/polish-synthpop-dr-4.json:36]: Same conflict with #38

File: run-inputs/polish-synthpop-dr-6.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    },
+    "parameters_sklearn_utility":
+    {
+        "enabled": true,
+        "input_columns": ["sex", "wkabint", "marital", "sport", "ls", "smoke"],
+        "label_column": "edu",
+        "test_train_ratio": 0.3,
+        "num_leaked_rows": 0
+    }
+}
```
Comment (gmingas) [inline run-inputs/polish-synthpop-dr-6.json]: Same conflict with #38

File: utility-metrics/sklearn_classifiers.py [hunk 4]
```diff
@@ -244,19 +361,38 @@ def main():
...
+                                 },
+
         # DecisionTreeClassifier: {"max_depth": 5},
-        # RandomForestClassifier: {"max_depth": 5, "n_estimators": 10, "max_features": 1},
         # MLPClassifier: {"alpha": 1, "max_iter": 5000},
         # AdaBoostClassifier: {},
-        GaussianNB: {},
-        QuadraticDiscriminantAnalysis: {}
+        #GaussianNB: {},
+        #QuadraticDiscriminantAnalysis: {}
```
Comment (gmingas) [inline utility-metrics/sklearn_classifiers.py]: Should these be uncommented?

**Review by gmingas — COMMENTED** [review]
Along with the comments on the code, I also get this error every time I run make, not sure if it specific to my environment, @kasra-hosseini have you seen this?:

```
Traceback (most recent call last):
  File "utility-metrics/sklearn_classifiers.py", line 38, in <module>
    report = importlib.import_module("utility-metrics.report")
  File "/Users/gmingas/opt/anaconda3/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 941, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'utility-metrics'
make: *** [synth-output/ctgan-example-0/sklearn_classifiers.json] Error 1

```

**Comment by kasra-hosseini** [comment]
> Along with the comments on the code, I also get this error every time I run make, not sure if it specific to my environment, @kasra-hosseini have you seen this?:
> 
> ```
> Traceback (most recent call last):
>   File "utility-metrics/sklearn_classifiers.py", line 38, in <module>
>     report = importlib.import_module("utility-metrics.report")
>   File "/Users/gmingas/opt/anaconda3/lib/python3.6/importlib/__init__.py", line 126, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "<frozen importlib._bootstrap>", line 994, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 971, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 941, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
>   File "<frozen importlib._bootstrap>", line 994, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 971, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
> ModuleNotFoundError: No module named 'utility-metrics'
> make: *** [synth-output/ctgan-example-0/sklearn_classifiers.json] Error 1
> ```

@gmingas Could you please check if you have: `QUIPP-pipeline/utility-metrics/__init__.py`?

File: run-inputs/not_used/polish-synthpop-dr-4.json (hunk 1 above)
Comment (kasra-hosseini) [reply run-inputs/not_used/polish-synthpop-dr-4.json]: Done. See [commit](https://github.com/alan-turing-institute/QUIPP-pipeline/pull/39/commits/89e7637603158d26eea68f8dda23378f3c60f52c)

File: utility-metrics/sklearn_classifiers.py (hunk 4 above)
Comment (kasra-hosseini) [reply utility-metrics/sklearn_classifiers.py]: In this PR, maybe we can keep the four classifiers that we have tested? I will add more classifiers in the next PR.

File: run-inputs/polish-synthpop-dr-4.json (hunk 2 above)
Comment (kasra-hosseini) [reply run-inputs/polish-synthpop-dr-4.json:36]: This should be solved now. See [commit](https://github.com/alan-turing-institute/QUIPP-pipeline/commit/1a87449c1976657382285c488c789a59352dc3e5)

File: run-inputs/polish-synthpop-dr-6.json (hunk 3 above)
Comment (kasra-hosseini) [reply run-inputs/polish-synthpop-dr-6.json]: This should be solved now. See [commit](https://github.com/alan-turing-institute/QUIPP-pipeline/commit/1a87449c1976657382285c488c789a59352dc3e5)

**Comment by gmingas** [comment]
> @gmingas Could you please check if you have: `QUIPP-pipeline/utility-metrics/__init__.py`?

Yes, I do. It is empty.

**Comment by kasra-hosseini** [comment]
@gmingas Can we please revisit this PR? It seems to me that the only remaining point is:

> I also get this error every time I run make, not sure if it specific to my environment, @kasra-hosseini have you seen this?:
> 
> ```
> Traceback (most recent call last):
>   File "utility-metrics/sklearn_classifiers.py", line 38, in <module>
>     report = importlib.import_module("utility-metrics.report")
>   File "/Users/gmingas/opt/anaconda3/lib/python3.6/importlib/__init__.py", line 126, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "<frozen importlib._bootstrap>", line 994, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 971, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 941, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
>   File "<frozen importlib._bootstrap>", line 994, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 971, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
> ModuleNotFoundError: No module named 'utility-metrics'
> make: *** [synth-output/ctgan-example-0/sklearn_classifiers.json] Error 1
> ```

I created a new conda env (python 3.6.11), installed the following packages:

```
pip install matplotlib
pip install pandas
pip install ctgan
pip install dython
```

the output of `pip freeze`:

```
certifi==2020.6.20
ctgan==0.2.1
cycler==0.10.0
dython==0.6.1
future==0.18.2
joblib==0.16.0
kiwisolver==1.2.0
matplotlib==3.3.1
numpy==1.19.1
pandas==0.25.3
Pillow==7.2.0
pyparsing==2.4.7
pyspark==2.4.3
python-dateutil==2.8.1
pytz==2020.1
scikit-learn==0.22.2.post1
scipy==1.5.2
seaborn==0.10.1
six==1.15.0
torch==1.6.0
torchvision==0.7.0
```

and everything seems to work. Could you please retry?

**Comment by gmingas** [comment]
I kept getting the same error and it was resolved when I changed the import from 
`report = importlib.import_module("utility-metrics.report")` 
to 
`report = importlib.import_module("report")`

**Pull request merged**